astropy/io/ascii/src/tokenizer.c

```diff
@@ -601,10 +601,15 @@ double str_to_double(tokenizer_t *self, char *str)
     {
         val = strtod(str, &tmp);
 
-        if (tmp == str || *tmp != 0)
+        if (tmp == str || *tmp != 0) {
+		    if (0 == strncmp(str, "nan", 3)) {
+				return NAN;
+		    }
```

Comment:
Should this code also work if the `xstrtod` is being used?  Since it seems like it also doesn't handle inf/nan.  I'll have a look at what Numpy does here...